enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+control-plane node. This will then ensure that the `router-default` deployment
+created by the Cluster Ingress Operator will always run on the single
+control-plane node, and as a result any `*.apps.<cluster>.<base>` DNS entries
+which originally pointed at the single control-plane node will remain correct
+even in the face of newly added worker nodes.
+
+In any other situations, the installer will set `IngressPlacement` to
+`Workers`, resulting in the same default behavior as before this enhancement,
+namely that `IngressController` pod replicas are scheduled on worker nodes and
+determined according to the `InfrastructureTopology`.
+
+In the future, when IPI-installed single control-plane node clusters in the
+cloud no longer provision a load-balancer by default, they would also benefit
+from having the installer set the `IngressPlacement` to `ControlPlane`.
+
+### Risks and Mitigations
+
+This should make no noticable difference on "regular" single control-plane node
+clusters which do not have any day-1 or day-2 worker nodes. The only difference
+for those clusters would be the `IngressController` targeting the "master" pool
+rather than the "worker" pool, but since the single control-plane node is
+already both in the "master" and "worker" pools, that should make no practical
+difference.
+
+I do not believe this enhancement has any security implications.
+
+## Design Details
+
+### Open Questions
+
+None that I can think of at the moment
+
+### Test Plan
+
+- Add unit tests to the Cluster Ingress Operator to make sure the
+IngressController resource is generated as exepected.
+
+- Add periodic nightly tests which install single-node in the cloud, add a few
+worker nodes to it, then run conformance tests to make sure we don't run into
+any problems not described in this enhancement.
+
+- Add periodic nightly tests which install a single-node "none"-platform cluster,
+add worker nodes to it, and check that ingress traffic still works as expected
+and recovers even after the `router-default` pod gets deleted and rescheduled.
+Make sure this is still true even after upgrades.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after the single control-plane node reboots by
+running conformance tests post-reboot.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after an upgrade by running conformance tests
+post-upgrade.
+
+TODO: Describe day-1-workers tests?
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+TBD
+
+#### Dev Preview -> Tech Preview
+
+TBD
+
+#### Tech Preview -> GA
+
+TBD
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+In the non-goals section it's mentioned that this enhancement does not apply to
+clusters which have been installed prior to the enhancement, so their upgrade
+is not discussed.
```

Comment:
I don't think we're going to be able to use that as a reason not to talk about upgrades here, because this change doesn't only apply to SNO deployments. All existing HA deployments will have the new field after an upgrade.

Reply:
Since these platforms already have IngressController CRs defined, this parameter should have no effect on them, as it only controls the defaulting behavior of such resources, and they already have those parameters set. 

You are correct that it has to be addressed. I will make sure it is

Reply:
EDIT: It seems that `replicas` is set explicitly, but the node selector is not. I'll give that some thought, if it could maybe cause problems

Reply:
Correct, the IngressController CR can be deleted.  If the user deletes the default IngressController CR, then the ingress operator recreates it after a short delay, and at that point, this enhancement would influence the defaults for the newly created CR.  That wouldn't happen during an upgrade or downgrade though, unless the user were deleting the CR during the upgrade/downgrade.

Reply:
It would not happen as part of an upgrade or downgrade. Pertinent here, in my mind however, is how a cluster that has been upgraded from a version that did not support ingress placement would behave.